etcd/v1alpha1/types_pacemakercluster.go

```diff
@@ -0,0 +1,549 @@
+	// The "PacemakerClusterNodeOnline" condition tracks whether the node is online.
+	// The "PacemakerClusterNodeNotInMaintenance" condition tracks whether the node is not in maintenance mode.
+	// The "PacemakerClusterNodeActive" condition tracks whether the node is active (not in standby mode).
+	// The "PacemakerClusterNodeReady" condition tracks whether the node is ready (not in a pending state).
+	// The "PacemakerClusterNodeClean" condition tracks whether the node is in a clean (status known) state.
+	// The "PacemakerClusterNodeMember" condition tracks whether the node is a member of the cluster.
+	// Each of these conditions is required.
+	// +listType=map
+	// +listMapKey=type
+	// +kubebuilder:validation:MaxItems=16
+	// +kubebuilder:validation:XValidation:rule="self.exists(c, c.type == 'PacemakerClusterNodeHealthy')",message="conditions must contain a condition of type PacemakerClusterNodeHealthy"
+	// +kubebuilder:validation:XValidation:rule="self.exists(c, c.type == 'PacemakerClusterNodeOnline')",message="conditions must contain a condition of type PacemakerClusterNodeOnline"
+	// +kubebuilder:validation:XValidation:rule="self.exists(c, c.type == 'PacemakerClusterNodeNotInMaintenance')",message="conditions must contain a condition of type PacemakerClusterNodeNotInMaintenance"
+	// +kubebuilder:validation:XValidation:rule="self.exists(c, c.type == 'PacemakerClusterNodeActive')",message="conditions must contain a condition of type PacemakerClusterNodeActive"
+	// +kubebuilder:validation:XValidation:rule="self.exists(c, c.type == 'PacemakerClusterNodeReady')",message="conditions must contain a condition of type PacemakerClusterNodeReady"
+	// +kubebuilder:validation:XValidation:rule="self.exists(c, c.type == 'PacemakerClusterNodeClean')",message="conditions must contain a condition of type PacemakerClusterNodeClean"
+	// +kubebuilder:validation:XValidation:rule="self.exists(c, c.type == 'PacemakerClusterNodeMember')",message="conditions must contain a condition of type PacemakerClusterNodeMember"
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+	// name is the name of the node. This is expected to match the Kubernetes node's name, which must be a lowercase
+	// RFC 1123 subdomain consisting of lowercase alphanumeric characters, '-' or '.', starting and ending with
+	// an alphanumeric character, and be at most 253 characters in length.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	// +kubebuilder:validation:XValidation:rule="!format.dns1123Subdomain().validate(self).hasValue()",message="name must be a lowercase RFC 1123 subdomain consisting of lowercase alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character"
+	// +required
+	Name string `json:"name,omitempty"`
+
+	// ipAddress is the canonical IPv4 or IPv6 address of the node. It must be a valid canonical global unicast IPv4
```

Comment:
Be careful with the word canonical here. This could mean a "canonical IP" as in an IP that is in canonical form, or it could mean the IP for the default route on the node depending on who you ask.

Also, which IP address is this? Nodes may have more than one, and quite often do. Does pacemaker care? Does it require/bind a specific address as part of its operations?

Have you considered making this a list to allow for multiple IP addresses? We have a bunch of examples of IP address lists that allow two addresses, where if there is more than one, each must be from a different IP family (4 vs 6)

Reply:
This turned out to  be a deeper rabbit hole than expected. I'll check in with the etcd guys, but it looks like we will want this to be list, but I want to understand how etcd endpoints are handled for dualstack clusters.